# Huỳnh Minh Huấn - 1612858

### Câu 1: Linear Regression Error

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import math

In [2]:
# Có thể sử dụng hàm find_N() => tìm ra được lowerBound của N thỏa câu 1
def find_N(sigma, d, val):
    return (d + 1)/(1 - 0.008 / math.pow(sigma, 2))
find_N(0.1, 8, 0.008)

44.99999999999996

In [3]:
def E_Ein(N, sigma, d):
    return math.pow(sigma, 2) * (1-(d + 1)/N)

In [4]:
sigma = 0.1
d = 8
print("N = 10 => Ed(Ein) =", E_Ein(10, sigma, d))
print("N = 25 => Ed(Ein) =", E_Ein(25, sigma, d))
print("N = 100 => Ed(Ein) =", E_Ein(100, sigma, d))
print("N = 500 => Ed(Ein) =", E_Ein(500, sigma, d))
print("N = 1000 => Ed(Ein) =", E_Ein(1000, sigma, d))

N = 10 => Ed(Ein) = 0.001
N = 25 => Ed(Ein) = 0.006400000000000001
N = 100 => Ed(Ein) = 0.009100000000000002
N = 500 => Ed(Ein) = 0.009820000000000002
N = 1000 => Ed(Ein) = 0.009910000000000002


### 1.

Ta thấy với N = 100 (giá trị nhỏ nhất), thì kỳ vọng $E_{in}\ge0.008$

**Chọn [c] 100.**

---

### Câu 2 - 3: Nonlinear Transforms

### 2.
Ta có: $h(\mathbf{x}) = sign(\tilde{w}_0 + \tilde{w}_1 x_1^2 + \tilde{w}_2 x_2^2)$

Khi $x_1$ siêu âm hoặc siêu dương thì $x_1^2$ sẽ dương và rất lớn, và  $\tilde{w}_1 x_1^2$ sẽ là đại lượng quyết định dấu của tổng trong công thức ở trên. Do $x_1^2$ dương nên $\tilde{w}_1$ sẽ quyết định dấu của $\tilde{w}_1 x_1^2$. Ta thấy dấu của tổng trong công thức trên sẽ âm khi |x1| lớn (x1 siêu âm hoặc siêu dương) - nhìn vào đồ thị. Do đó $\tilde{w}_1 < 0$.

Tương tự, khi $x_2$ siêu âm hoặc siêu dương thì $x_2^2$ sẽ dương và rất lớn, và  $\tilde{w}_2 x_2^2$ sẽ là đại lượng quyết định dấu của tổng trong công thức ở trên. Do $x_2^2$ dương nên $\tilde{w}_2$ sẽ quyết định dấu của $\tilde{w}_2 x_2^2$. Ta thấy dấu của tổng trong công thức trên sẽ dương khi |x2| lớn (x2 siêu âm hoặc siêu dương) - nhìn vào đồ thị. Do đó $\tilde{w}_2 > 0$.

Vậy: $\tilde{w}_1 < 0,\tilde{w}_2 > 0$.

**Chọn [d] $\tilde{w}_1 < 0,\tilde{w}_2 > 0$.**

---

### 3.
Sử dụng công thức về $d_{VC}$ của mô hình phân lớp tuyến tính (Perceptron) trong không gian $\mathcal{Z}$ ở slide 4, lecture 9 - the linear model II ${d_{VC}\leq d^*+ 1}$ với $d^*+ 1$ là số chiều của w* trong không gian $\mathcal{Z}$ (không gian khi chuyển áp dụng non-linear transform).

Ta có: d* + 1 = 15, suy ra $d_{VC} \leq 15$.

Do đó: giá trị nhỏ nhất trong các đáp án mà không nhỏ hơn $d_{VC}$ là: 15.

**Chọn [c] 15.**

---

### Câu 4 - 7: Gradient Descent

### 4.

${\frac{\partial E}{\partial u} = \frac{\partial(ue^v-2ve^{-u})^2}{\partial u}}$

${\frac{\partial E}{\partial u} = 2(ue^v-2ve^{-u})(ue^v-2ve^{-u})'}$

**Kí hiệu:** $(ue^v-2ve^{-u})' = \frac{\partial(ue^v-2ve^{-u})}{\partial u}$ do đang xét đạo hàm riêng theo u nên em ghi vậy cho gọn.

Ta có:

- $\frac{\partial(ue^v)}{\partial u} = e^v$
- $\frac{\partial(-2ve^{-u})}{\partial u} = -2*(-u)'ve^{-u} = 2ve^{-u}$

Do đó: ${\frac{\partial E}{\partial u} = 2(ue^v-2ve^{-u})(e^v + 2ve^{-u})}$

**Chọn [e] ${2(e^v + 2ve^{-u})(ue^v-2ve^{-u})}$.**

---

#### Hàm Gradient tính gradient của E và Hàm Gradient_descent chạy thuật toán GD

In [5]:
def Gradient(point):
    """
    input: point(u, v)
    output: [du, dv]
    """
    du = 2 * (np.exp(point[1]) + 2*point[1]*np.exp(-point[0]))*(point[0]*np.exp(point[1])-2*point[1]*np.exp(-point[0]))
    dv = 2 * (point[0]*np.exp(point[1])-2*point[1]*np.exp(-point[0])) * (point[0]*np.exp(point[1])-2*np.exp(-point[0]))
    return np.array([du, dv])

def Gradient_Descent(eta, threshold):
    """
    input: 
        eta - leanring rate
        threshold - ngưỡng
    ouput:
        iterations - số lần lặp
        point - điểm cực tiểu cục bộ tìm được
    """
    point = np.array([1, 1])
    Error = math.pow(1*np.exp(1) - 2*1*np.exp(-1), 2)
    iterations = 0
    while Error >= threshold:
        iterations += 1
        point = point - eta * Gradient(point)
        Error = math.pow(point[0]*np.exp(point[1]) - 2*point[1]*np.exp(-point[0]), 2)
    return iterations, point

In [6]:
def main_5_6(eta, threshold):
    #eta = 0.1
    #threshold = 1e-14
    iterations, point = Gradient_Descent(eta, threshold)
    print("Iterations =",iterations)
    print("Point =",point)
main_5_6(eta = 0.1, threshold = 1e-14)
#print(Gradient([13.69542993, 7.86079446]))

Iterations = 10
Point = [0.04473629 0.02395871]


### 5.

Chạy hàm main_5_6(eta = 0.1, threshold = 1e-14), ta được kết quả: **Iterations = 10**.

**Chọn [d] 10.**

---

### 6.

Chạy hàm main_5_6(eta = 0.1, threshold = 1e-14), ta được kết quả: **Point = [0.04473629 0.02395871]**. Gần nhất với đáp án **(0.045, 0.024)**.

**Chọn [e] (0.045,0.024).**

---

In [7]:
def main_7(eta, N):
    #eta = 0.1
    #N = 15
    grad = np.array([1, 1])
    for i in range(N):
        du = 2 * (np.exp(grad[1]) + 2*grad[1]*np.exp(-grad[0]))*(grad[0]*np.exp(grad[1])-2*grad[1]*np.exp(-grad[0]))
        grad = grad - eta * np.array([du, 0])
        dv = 2 * (grad[0]*np.exp(grad[1])-2*grad[1]*np.exp(-grad[0])) * (grad[0]*np.exp(grad[1])-2*np.exp(-grad[0]))
        grad = grad - eta * np.array([0, dv])
    Error = math.pow(grad[0]*np.exp(grad[1]) - 2*grad[1]*np.exp(-grad[0]), 2)
    print("Error = ", Error)
main_7(eta = 0.1, N = 15)

Error =  0.13981379199615315


### 7.

Chạy hàm main_7(eta = 0.1, N = 15), ta được kết quả: **Error =  0.13981379199615315**. Gần nhất với đáp án **0.1**.

**Chọn [a] $10^{-1}$.**

---

### Câu 8 - 9: Logistic Regression

Hàm phát sinh ra `target_w`, véc-tơ tham số của $f$

In [8]:
def generate_target_w():
    """
    Generates target_w from two random, uniformly distributed points in [-1, 1] x [-1, 1].
    
    Returns
    -------
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    """
    # Generate two points from a uniform distribution over [-1, 1]x[-1, 1]
    p1 = np.random.uniform(-1, 1, 2)
    p2 = np.random.uniform(-1, 1, 2)
    # Compute the target W from these two points
    target_w = np.array([p1[1]*p2[0] - p1[0]*p2[1], p2[1] - p1[1], p1[0] - p2[0]]).reshape((-1, 1))
    
    return target_w

Hàm phát sinh ra tập dữ liệu

In [9]:
def generate_data(N, target_w):
    """
    Generates a data set by generating random inputs and then using target_w to generate the 
    corresponding outputs.
    
    Parameters
    ----------
    N : int
        The number of examples.
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    
    Returns
    -------
    X : numpy array, shape (N, 3)
        The matrix of input vectors (each row corresponds to an input vector); the first column of 
        this matrix is all ones.
    Y : numpy array, shape (N, 1)
        The vector of outputs.        
    """
    bad_data = True # `bad_data = True` means: data contain points on the target line 
                    # (this rarely happens, but just to be careful)
                    # -> y's of these points = 0 (with np.sign); 
                    #    we don't want this (y's of data must be -1 or 1)
                    # -> re-generate data until `bad_data = False`
    
    while bad_data == True:
        X = np.random.uniform(-1, 1, (N, 2))
        X = np.hstack((np.ones((N, 1)), X)) # Add 'ones' column
        Y = np.sign(np.dot(X, target_w))
        if (0 not in Y): # Good data
            bad_data = False
    
    return X, Y

Hàm chạy thuật toán Stochastic Gradient Descent

In [10]:
def SGD(X, Y, lr = 0.01):
    """
    input:
        X : numpy array, shape (N, 3)
        Y : numpy array, shape (N, 1)    
        lr: float
    output:
        w : numpy array, shape(3)
        n_epochs: int
    """
    N, sizeW = X.shape
    w = np.zeros([sizeW])
    #print(w)
    n_epochs = 0
    epochs = []
    E_ins = []
    while True:
        pre_w = np.copy(w)
        rand_idx = np.random.permutation(N)
        for id in rand_idx:
            # do something
            gradient = (-Y[id] * X[id,:])/(1 + np.exp(Y[id]*np.dot(w.T, X[id,:])))
            w = w - lr * gradient 
        #E_in = 0
        #for j in range(100):
        #    E_in += np.log(1 + np.exp(-Y[j]*np.dot(w.T, X[j])))
        n_epochs += 1
        #epochs.append(n_epochs)
        #E_ins.append(E_in)
        if np.linalg.norm(pre_w - w) < 0.01:
            break
    #plt.plot(epochs, E_ins)
    #plt.show()
    return w.reshape(3), n_epochs

In [11]:
target_w = generate_target_w()
X, Y = generate_data(100, target_w)
w, n_epochs = SGD(X, Y, lr = 0.01)
print(w)
print(n_epochs)

[0.87429193 8.28347753 4.20687108]
332


In [12]:
def main_8_9():
    N = 100 # data training points
    epochs = []
    Eouts = []
    average_Eout = 0
    # Repeat the experiment for 100 runs with different targets and take the average.
    for i in range(100): 
        # print(i)
        target_w = generate_target_w()
        X, Y = generate_data(N, target_w)
        w, n_epochs = SGD(X, Y, lr = 0.01)
        #Eout = 0
        X_new, Y_new = generate_data(N, target_w) # generate new data kích thước N = 1000 để tính Eout
        # Tính Eout theo công thức cross entropy với new Data
        #for j in range(1000):
        #    Eout += np.log(1 + np.exp(-Y_new[j]*np.dot(w.T, X_new[j])))
        #Eouts.append(Eout / 1000)
        #Eouts.append(np.mean(np.log(1 + np.exp(-Y_new*np.dot(X_new, w)[:, None]))))
        average_Eout += np.mean(np.log(1 + np.exp(-Y_new*np.dot(X_new, w)[:, None]))) / N
        epochs.append(n_epochs)
    print("average_Eout = ",np.mean(average_Eout))
    print("average_epochs = ",np.mean(epochs))
%time main_8_9()

average_Eout =  0.10189830919779982
average_epochs =  340.11
Wall time: 1min 26s


### 8.

Chạy hàm main_8_9(), ta được kết quả average_Eout xấp xỉ **0.101...** gần nhất với 0.100.

**Chọn [d] 0.100.**

---

### 9.
Chạy hàm main_8_9(), ta được kết quả average_epochs xấp xỉ gần nhất 350.

**Chọn [a] 350.**

---

### Câu 10: PLA vs SGD

### 10.
- PLA: 
    - Với mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ được phân lớp đúng thì $\mathbf{w}$ không thay đổi.
    - Với mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ bị phân lớp sai thì cập nhật: $\mathbf{w} \leftarrow \mathbf{w} + y^{(n)}\mathbf{x}^{(n)}$.

- SGD: với mỗi mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ thì đều cập nhật $\mathbf{w} \leftarrow \mathbf{w} - \alpha \times \mathbf{\nabla_wE_{in}}$.

Xét $\alpha=1$:
SGD mỗi mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ cập nhật: $\mathbf{w} \leftarrow \mathbf{w} -\mathbf{\nabla_wE_{in}}$.

Ta có $E_{in}=\frac{1}{N}\sum_{n=1}^Ne_n$. Suy ra: $\nabla_wE_{in} = \nabla_w\frac{1}{N}\sum_{n=1}^Ne_n = \frac{1}{N}\sum_{n=1}^N\nabla_w e_n$.

Xét: $e_n = -min(0, y^{(n)}w^Tx^{(n)})$ 

- Với mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ được phân lớp đúng $e_n = 0$ thì $\nabla_w e_n = 0$. Suy ra: $\nabla_wE_{in} = 0$, $\mathbf{w} \leftarrow \mathbf{w} - 0 = \mathbf{w}$

- Với mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ bị phân lớp sai $e_n = -y^{(n)}w^Tx^{(n)}$ thì $\nabla_w e_n = -y^{(n)}x^{(n)}$. Suy ra: $\nabla_wE_{in} = \frac{1}{N}-Ny^{(n)}x^{(n)} = -y^{(n)}x^{(n)}$, $\mathbf{w} \leftarrow \mathbf{w} -  (-y^{(n)}x^{(n)}) = \mathbf{w} + y^{(n)}x^{(n)}$

Thỏa PLA nên chọn [e]

**Chọn [e] $e_n(w)=-min(0,y_nw^Tx_n)$.**

---